File: 239-sliding-window-maximum/239-sliding-window-maximum.py
class Solution:
    def maxSlidingWindow(self, nums: List[int], k: int) -> List[int]:
        
        if not nums: return []
        res = []
        dq = deque()  # store index
        for i in range(len(nums)):
            if dq and dq[0]<i-k+1:  # out of the window
                dq.popleft()
            while dq and nums[dq[-1]]<nums[i]:  # remove impossible candidate
                dq.pop()
            dq.append(i)
            if i>k-2:
                res.append(nums[dq[0]])
        return res

File: 239-sliding-window-maximum/239-sliding-window-maximum.cpp
class Solution {
public:
    vector<int> maxSlidingWindow(vector<int>& nums, int k) {
        
        vector<int> ret; 
        deque<int> deq; 
        
        for( int i =0; i < nums.size(); ++i){
            
            if(!deq.empty() && i-k >=0 && deq.front()==nums[i-k]){
                deq.pop_front();
            }
            while(!deq.empty() && deq.back() < nums[i] ){
                deq.pop_back();
            }
            deq.push_back(nums[i]);
            
            if (i>=k-1) ret.push_back(deq.front());
        }
        return ret; 
        
        
    }
};